atomic_reactor/plugins/post_koji_upload.py

```
@@ -0,0 +1,643 @@
+        auth_info = {
+            "proxyuser": self.koji_proxy_user,
+            "ssl_certs_dir": self.koji_ssl_certs_dir,
+            "krb_principal": str(self.koji_principal),
+            "krb_keytab": str(self.koji_keytab)
+        }
+        return create_koji_session(str(self.kojihub), auth_info)
+
+    def run(self):
+        """
+        Run the plugin.
+        """
+
+        if ((self.koji_principal and not self.koji_keytab) or
+                (self.koji_keytab and not self.koji_principal)):
+            raise RuntimeError("specify both koji_principal and koji_keytab "
+                               "or neither")
+
+        # Only run if the build was successful
+        if self.workflow.build_process_failed:
+            self.log.info("Not promoting failed build to koji")
+            return
+
+        koji_metadata, output_files = self.get_metadata()
+
+        try:
+            session = self.login()
+            server_dir = self.get_upload_server_dir()
+            for output in output_files:
+                if output.file:
+                    self.upload_file(session, output, server_dir)
+        finally:
+            for output in output_files:
+                if output.file:
+                    output.file.close()
+
+        md_fragment = "{}-md".format(get_build_json()['metadata']['name'])
+        md_fragment_key = 'metadata.json'
+        cm_data = {md_fragment_key: koji_metadata}
+        annotations = {
+            "metadata_fragment": "configmap/" + md_fragment,
+            "metadata_fragment_key": md_fragment_key
+        }
+
+        try:
+            self.osbs.create_config_map(md_fragment, cm_data)
+        except OsbsException:
+            self.log.debug("metadata: %r", koji_metadata)
+            self.log.debug("annotations: %r", annotations)
+            raise
+        except AttributeError as e:
+            self.log.debug("missing attribute: %s", e.strerror)
```

It's probably better to not catch this exception and just let it propagate normally.